### PR TITLE
fix: set the initial repl cursor pos, fixes #8943

### LINKED ISSUE #8943: commandline --insert not inserting at cursor position any more
### Describe the bug

I'm using `commandline --insert` in some of my keybindings and this worked, but since the upgrade to version 0.78.0, this does not work anymore and the insertion is always done at the front of the commandline. 

In 0.77.1 this worked fine.

### How to reproduce

1. Create a default nushell configuration
2. Add the following keybing:

```
{
                name: bar_test
                modifier: control
                keycode: char_t
                mode: [emacs, vi_insert, vi_normal]
                event: {
                  send: executehostcommand             
                  cmd: `commandline --insert bar`
                }
    }
```
3. (Re)start nushell to load the new config
4. Type `foo`
5. Press `CTRL-T`
6. Result: `barfoo`

### Expected behavior

I expected 'commandline --insert' to insert the string at the cursor position, not the start of the line.
So in the example above, I expected to see `foobar` instead of `barfoo`

### Screenshots

_No response_

### Configuration

| key                | value                                                            |
| ------------------ | ---------------------------------------------------------------- |
| version            | 0.78.0                                                           |
| branch             | makepkg                                                          |
| commit_hash        | 2ec20286371f1d90a77fdd5d82818c6c5851f090                         |
| build_os           | linux-x86_64                                                     |
| build_target       | x86_64-unknown-linux-gnu                                         |
| rust_version       | rustc 1.68.2 (9eb3afe9e 2023-03-27) (Arch Linux rust 1:1.68.2-1) |
| cargo_version      | cargo 1.68.2                                                     |
| build_time         | 2023-04-05 02:09:35 +00:00                                       |
| build_rust_channel | release                                                          |
| features           | default, zip                                                     |
| installed_plugins  |                                                                  |

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by stevenxxiu**
# Description
Set the initial repl cursor pos, so running `commandline --insert` inserts at the current cursor position of the input buffer.

# User-Facing Changes

# Tests + Formatting

# After Submitting

### CODE REVIEW

**Comment by sholderbach** [comment]
This fix feels like a bit of a band-aid to me. We have robust commands for insertion/modification directly in the `reedline::LineBuffer`. Separately messing with the `String` and the insertion point as a `usize` in nushell is a point-of-failure in my view. So I am not a fan of exposing the implementation details with nushell/reedline#573

The chain of reading and writing the mutexed `repl_buffer_state` and `repl_cursor_pos` is hard to follow and easy to mess up with further changes.

Before #8207 the valid actions were explicitly encoded in an enum which is like the first step to isolating parts that are easy to validate and read.

**Comment by stevenxxiu** [comment]
The point of `commandline` is to enable both getting/setting the cursor and command text, so it's hard to isolate things further. The only point of failure I can see here is if the cursor isn't on a grapheme boundary. I'm open to suggestions.

Before https://github.com/nushell/nushell/pull/8207 the actions were encoded in an enum, but it lacked setting the cursor position. I don't really see why an enum is needed though, given that `commandline` allows modifying both variables in every way.

**Comment by sholderbach** [comment]
Sorry for the rant, the enum of old is probably not the best solution.

We can land this fix for now but should clean that up.

Modifying the cursor position only makes sense when taking the buffer into account and vice versa, to reflect the grapheme boundaries as well as the overall length. 

As they only are validly changed together, they (`repl_buffer_state` and `repl_cursor_pos` ) shouldn't be behind two separate mutexes which expresses that they can change independently. 

Testing this fix I noticed that the read cursor position is still in bytes while the written one is graphemes.

```
~/nushell> commandline aä; commandline -c 2; commandline -c
3
~/nushell> aä 
```
